### ISSUE TITLE
Initial P4 support tracking issue

### ISSUE THREAD

**Issue by MabezDev** [esp-rs/esp-hal#2285]
Whilst we don't intend to do a full chip bring-up until the mass production version is released, we would like to get a minimal chip bring-up to aid future developments.

- [x] PAC support
    - We have a PAC already in `esp-pacs`, will require patching I'm sure but at least for now we can probably consider this subtask complete
- [x] Linker and required `soc` module features
- [ ] Tooling(?)
    - https://github.com/esp-rs/espflash/pull/686
    - https://github.com/probe-rs/probe-rs/pull/2047
- [ ] Uart driver - https://github.com/esp-rs/esp-hal/pull/2466
- [x] HIL runner

**Comment by jessebraham** [comment]
I again have a very basic application running on device, calling only `esp_println::println!()`.

There is a caveat to this, in that the application works on the `v1.0` devkit but neither the `v1.1` not `v1.2` version; I have some ideas with regards to this, but will require more experimentation. At least for the time being, I'm able to move forward using the old devkit.

Unfortunately I think getting a UART example running will take a bit more time than expected, as there is quite tight coupling between the interrupt support and various drivers. Given that the ESP32-P4 uses yet another interrupt controller (CLIC) this will take some time to sort out. Hoping to have a hacky PoC by end of week or early next week, though.

I will continue to update this issue as I make progress.

**Comment by bjoernQ** [comment]
I remember we had interrupts and gpio working in January or February already

**Comment by jessebraham** [comment]
> I remember we had interrupts and gpio working in January or February already

Yes but there have been a number of changes to the HAL since January, and it's not as simple as copy-pasting the previous code in unfortunately. I do not think this will take a ton of time to be clear, but it will take some time to understand the changes which have been made and adapt the previous implementation to match.

**Comment by playfulFence** [comment]
Working on this right now. I'm not sure how far I'll get, but at least I want to adapt all the changes to the current state of a driver.

**Comment by jessebraham** [comment]
@playfulFence Is the P4 HIL runner in a functional state? Can we tick off that task?

**Comment by bugadani** [comment]
Define functional, we have a raspberry that boots. We don't have anything on it, let alone flashing support in probe-rs or GHA workflows in this repo. What's the definition of done of that checkbox?

**Comment by jessebraham** [comment]
> Define functional, we have a raspberry that boots. We don't have anything on it, let alone flashing support in probe-rs or GHA workflows in this repo. What's the definition of done of that checkbox?

Been awhile since I opened this, but given that there is a separate task for tooling (which links to the `probe-rs` issue), I think just having the runner set up physically and having the GitHub runner configured is adequate.

**Comment by MabezDev** [comment]
Dropping this back to the current milestone, there's a chance we can complete it in time, but if not we'll reassess what to do in the next planning meeting.

**Comment by MabezDev** [comment]
@playfulFence has been working hard to get interrupt support working. It's _nearly_ there, and works correctly _without_ vectoring the CPU interrupts (this is the RISCV interrupt controller vectoring, different from vectoring the peripheral interrupts). The P4 is quite different to the other esp's, and requires a number of changes in esp-riscv-rt (based on esp-idf, we didn't actually get this working just yet).

I think we'll want to revisit this support once https://github.com/esp-rs/esp-hal/issues/2390 has landed, as upstream riscv-rt has better support for this and we should be directing our efforts there anyway.

We'll leave the PR open https://github.com/esp-rs/esp-hal/pull/2466 as draft, and we can occasionally rebase if anyone would like to try out running some pure Rust code on a early P4 devkit.

**Comment by jamesmunns** [comment]
Sounding off that I'm interested in helping, and I'll start pointing people here when they ask.

**Comment by rmsyn** [comment]
> Sounding off that I'm interested in helping, and I'll start pointing people here when they ask.

Same, happy to help out with testing and dev.

**Comment by SergioGasquez** [comment]
Thanks for offering help @jamesmunns, @rmsyn! Ill get back to you next week, once Scott is back from vacation, with some useful information about how/when we plan to support the P4!

**Comment by SergioGasquez** [comment]
Hi @jamesmunns, @rmsyn, sorry for the late reply, I went on vacation. Here is our current chip support plan: #3962 with updated p4 plans, if you have any questions, just let me know

**Comment by jamesmunns** [comment]
Only questions are when we might expect eco5 silicon to become available, and maybe HOW different it is expected to be from eco4, like, "a few things you could cfg feature gate" or "lmao totally different entirely don't even try".

Like, it might be fun to start with eco4 hardware, but if it's going to be a total throw away, less so.

This also is going to be really frustrating for people who bought (eco4) boards now and want to use them later when there's driver support, but I know this wasn't your decision.

**Comment by bugadani** [comment]
@jamesmunns I think the best indicator for changes is esp-idf: they have defined a separate hardware version, with a different register abstraction for ECO5. The analogy would be two somewhat different PAC crates for us.

**Comment by rmsyn** [comment]
> The analogy would be two somewhat different PAC crates for us.

So, any HAL work would need to potentially use conditional compilation to handle register naming differences in the PACs?

**Comment by bugadani** [comment]
It's not a matter of naming, it's a matter of different hardware behind the PACs. There are more differences between the devices than their register layout, though - different ROM code, different custom instruction set, and perhaps more. We essentially need to treat the old P4 as a completely separate MCU.

**Comment by vliberio** [comment]
I apologize for any inconvenience, but I was very curious whether this was still the case given the seeming mass-production volume of ESP32-P4NRW32 and other chips. Are these the aforementioned Eco5 P4s? Is P4 support still considered blocked as https://github.com/esp-rs/esp-hal/issues/3962 indicates?

Sorry for bothering you all, but I was wondering whether it would be a bad idea to begin designing hardware using the P4 with the assumption or hope that Rust support is under development? Thanks.

**Comment by bugadani** [comment]
ECO5 chips are publicly called "Chip revision v3.0". You can probably contact your supplier for information about the revision they are selling. Here is how to identify the chip revision (although it doesn't include the newer ones yet, v3.0 has a marking of `F`): https://docs.espressif.com/projects/esp-chip-errata/en/latest/esp32p4/01-chip-identification/index.html#chip-marking

**Comment by florianL21** [comment]
Looking at #3857 this seems to be unblocked right?
I have a P4 dev board in my hands right now and would be happy to help out with anything I can to get some basic support for the P4

**Comment by rdelfin** [comment]
@florianL21 have you identified what work is missing? I'd also like to help too to try it out!

**Comment by florianL21** [comment]
@rdelfin I already did a bunch of work but it's not yet compiling. I am actively working on it though.
You can see my current changes here: https://github.com/florianL21/esp-hal/commit/cce6529f6cec3dd76ed358f0ee6be24a3851dc88

And the branch I am working on is also in my fork and I will update it whenever I make some progress.

It's possible that I am doing some things incorrectly but I have to reverse enginner stuff based on other processors since I didn't find any documentation on how to add support for a new chip

**Comment by bugadani** [comment]
> It's possible that I am doing some things incorrectly but I have to reverse enginner stuff based on other processors since I didn't find any documentation on how to add support for a new chip

You're attempting to add support for a chip that isn't in mass production yet, and is significantly different from any other that we currently support. All the concepts are in flux, the P4 poses new challenges, I'm not sure if our current solutions are even compatible with the chip. We're not providing documentation for something that we don't really know how to do.

Having said that, your branch looks promising.

**Comment by florianL21** [comment]
Yes I noticed that sometimes the concepts you have so far do not really fit all that well to that chips architecture. 
Like the fact that there seem to be the HP_SYS and LP_SYS both responsible for enabling different peripherals. There is no single SYSTEM which does everything.

I have decided to for now focus only on the HP_SYS side to maybe get something running.
I fully understand that it may need a lot of additional work to get stuff to a production state. I'm happy if I would be able to use it for some projects. And even if it never ends up getting merged maybe it can at least be useful to someone when implementing the official support.

As for the production silicon: as far as I read into the topic and from what I could find also from component suppliers like LCSC the chip seems to be at least in stock in large-ish quantities and there is no real mention of these chips being pre-production samples. But it may also be just LCSC being not the best at communicating such things.
But there also seem to be a growing number of dev boards avaliable.
And esp-idf support seems quite complete.

So personally I am a bit sad that I cannot use rust with it yet, but I do understand why you don't want to waste your time supporting an eval chip.

**Comment by vliberio** [comment]
Yeah I was (and maybe am) still a little confused myself on P4's mass production & distribution status and contacted Espressif (sales@espressif.com) last month and got a couple emails back. I'll put the relevant responses here in case it helps anyone.

My Questions:
```
1. Are SoCs such as the ESP32-P4NRW32 the full, supported, production-ready ESP32-P4 product? The 'Marketing Status' in the Product Selector indicates these are 'Mass Production', but the datasheet from the same Product Selector page indicates it is 'Pre-release v0.6' and has 'PRELIMINARY' watermarked over each page? Are these just a sample / pre-release variant of the ESP32-P4?

2. If these are these are fully supported production ESP32-P4 variants, is there an official vendor / distributor for the ESP32-P4 that we could obtain a small batch (say 50-100?) for prototypes? The only listing we've so far seen was on LCSC (https://www.lcsc.com/product-detail/C22387510.html). Is LCSC an authorized distributor or is there one that offers the ESP32-P4 in the US?

3. Will the ESP32-P4 receive limited or full support for the official Rust SDK in the foreseeable future like with the other boards? (https://developer.espressif.com/blog/2025/10/esp-hal-1/)
```

Response #1 Snippet:
```
For question 2, LCSC (https://www.lcsc.com/product-detail/C22387510.html) is not one of  our authorized distributors. You can obtain ESP32-P4 samples through Espressif’s official store on AliExpress via the following link: https://www.aliexpress.com/item/1005008985298657.html 

Regarding questions 1 and 3, we will provide you with further responses as soon as possible. Thank you.
```

Response #2 Snippet:
```
1.	ESP32-P4NRW32 is a fully supported, production-ready ESP32-P4 SoC, and the “Mass Production” status in the Product Selector is correct.
Sorry, we still have some internal details to finalize. Once these are completed, we will release the v1.0 version of the datasheet as soon as possible.
2.	Our business colleagues have already provided a response in the previous email. Please refer to it.
3.	I have already consulted the relevant colleagues, and if there is any reply, I will immediately get back to you via email.
```

Personally, I was waiting for the datasheet to get finalized before proceeding further.

**Comment by H01001000** [comment]
The data sheet for p4 rev 1.3 is now revision 1.0
Tho they add NRND to ESP32-P4NRW32

**Comment by MabezDev** [comment]
Hi all,

Thanks for the interest here, we're also excited about P4 support! I'd just like to clarify our position on the NRND P4 variants around. We don't want, nor will we accept support for these variants. 

I know many of you have devkits already (as do I), but at a silicon level these are very different chips, and would put an enormous maintenance burden on us supporting both version. At this time I can only recommend ESP-IDF for development with these NRND chips.

We *will* support the full mass production ESP32-P4 in esp-hal, rest assured - though bare in mind, chip support is a low-ish prio task for us.

**Comment by vvreutskiy** [comment]
[Initial ESP32-P4 support](https://github.com/probe-rs/probe-rs/pull/2047) is closed by [esp32p4: add initial support](https://github.com/probe-rs/probe-rs/pull/3649) , so looks like tooling is mostly done

**Comment by pmnxis** [comment]
Current left jobs from my knowledge.
- #5400 is coming soon to main branch. <- This is initial works only considered P4X (Revision 3.0+)
- Some of improvement : #5555 #5523 #5559 
- **SLEEP** : #5564 
- **SPI Master** (just test and ensure it's working well with several situation, I didn't tested it)
- **CLIC** v1.x issue compare to C5/C61's CLIC v2.x (maybe PAC or separation required later , also espressif should ensure CLIC version of mass production revision) : https://github.com/esp-rs/esp-hal/pull/5400#pullrequestreview-4217848881
- **RMII/MII PHY/MAC** and actual working model with simple tcp or ping stuff (I failed at link with DHCP with #5400) : #5546, #5556 
- **USB HS/FS** and better interface considering single HS and single FS but more option compare to other ESP series) : #5558 #5560 
- **LP** Core stuff
- misc (i2c/spi/clock/rtc_cntl maybe need more investigation)
- Tyding to future TRM (current May4th it's 0.6)
- CD/CAM, MIPI CSI/DSI, ISP, PPA
- Multi-Core and utilize example (SMP/AMP?) : #5535

**Comment by MabezDev** [comment]
I'd consider the P4 supported at this point, closing.

**Comment by Avamander** [comment]
@MabezDev What about the relatively widespread ESP32-P4 rev1/rev2 chips being supported?

I'm pretty sure the current HAL doesn't account for the differences and it's not a feature one could toggle. Things such as the non-standard CLIC, different ROM, clock speed, etc.

**Comment by bugadani** [comment]
As stated in https://github.com/esp-rs/esp-hal/issues/3962, chips before v3 will not be supported by us.
